# **Análise do Mercado Imobiliário em Goiania**

In [4]:
# importacao de bibliotecas que usaremos
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import time

In [5]:
# leitura do dataset em csv
gyn = pd.read_csv('2021-08-05-all.csv')

In [6]:
# dataset overview
gyn.head()

,DATE,PRICE,ADDRESS,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU,TIPO
0,2021-08-05 16:34:17.725797-03:00,Sob consulta,"Rua João de Abreu, Setor Oeste",222 - 485 m²,3 - 4,2 - 4,3 - 5,NaN,NaN,apartamentos
1,2021-08-05 16:34:17.726263-03:00,R$ 430.000,"Rua R 7, Setor Oeste",66 - 68 m²,2,2,2 - 3,NaN,NaN,apartamentos
2,2021-08-05 16:34:17.726628-03:00,R$ 345.000,"Rua 7, Setor Oeste",34 - 76 m²,1 - 2,1,1,NaN,NaN,apartamentos
3,2021-08-05 16:34:17.726966-03:00,R$ 420.000,"Setor Bueno, Goiânia",89 m²,3,1,2,NaN,NaN,apartamentos
4,2021-08-05 16:34:17.727328-03:00,R$ 1.018.000,"Setor Bueno, Goiânia",175 m²,4,5,4,R$ 650,R$ 2.000,apartamentos


In [7]:
# informacao das colunas
gyn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22110 entries, 0 to 22109
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   DATE            22110 non-null  object
 1   PRICE           22110 non-null  object
 2   ADDRESS         22107 non-null  object
 3   AREAS           22072 non-null  object
 4   BEDROOMS        20665 non-null  object
 5   PARKING-SPACES  19470 non-null  object
 6   BATHROOMS       20681 non-null  object
 7   CONDOMÍNIO      11358 non-null  object
 8   IPTU            6669 non-null   object
 9   TIPO            22110 non-null  object
dtypes: object(10)
memory usage: 1.7+ MB


In [8]:
# formato das colunas
gyn.shape

(22110, 10)

In [9]:
# descricao acerca de todos os dados do dataset
gyn.describe()

,DATE,PRICE,ADDRESS,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU,TIPO
count,22110,22110,22107,22072,20665,19470,20681,11358,6669,22110
unique,22110,2135,3931,1087,21,35,28,444,629,11
top,2021-08-05 16:45:57.316281-03:00,R$ 450.000,"Setor Bueno, Goiânia",64 m²,3,2,2,R$ 500,R$ 1.500,apartamentos
freq,1,321,1206,296,10791,7715,6019,785,325,10656


In [10]:
# como estao organizados os dados
gyn.dtypes

,0
DATE,object
PRICE,object
ADDRESS,object
AREAS,object
BEDROOMS,object
PARKING-SPACES,object
BATHROOMS,object
CONDOMÍNIO,object
IPTU,object
TIPO,object


In [11]:
# verificacao de dados duplicados
gyn.duplicated().sum()

np.int64(0)

In [12]:
print('Numero de linhas e colunas:', gyn.shape)
print('Lista dos nomes das colunas:', gyn.columns.tolist())
print('Tipos inferidos:', gyn.dtypes)

Numero de linhas e colunas: (22110, 10)
Lista dos nomes das colunas: ['DATE', 'PRICE', 'ADDRESS', 'AREAS', 'BEDROOMS', 'PARKING-SPACES', 'BATHROOMS', 'CONDOMÍNIO', 'IPTU', 'TIPO']
Tipos inferidos: DATE              object
PRICE             object
ADDRESS           object
AREAS             object
BEDROOMS          object
PARKING-SPACES    object
BATHROOMS         object
CONDOMÍNIO        object
IPTU              object
TIPO              object
dtype: object


In [13]:
print("Tipo:", type(gyn))
print("Dimensão de df:", gyn.shape)

Tipo: <class 'pandas.core.frame.DataFrame'>
Dimensão de df: (22110, 10)


In [14]:
# Overview Aprofundada - Dataset
from IPython.display import display

print("5 primeiras linhas:")
display(gyn.head())

print("5 últimas linhas:")
display(gyn.tail())

print("5 linhas aleatórias:")
display(gyn.sample(n=5, random_state=42))

5 primeiras linhas:


,DATE,PRICE,ADDRESS,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU,TIPO
0,2021-08-05 16:34:17.725797-03:00,Sob consulta,"Rua João de Abreu, Setor Oeste",222 - 485 m²,3 - 4,2 - 4,3 - 5,NaN,NaN,apartamentos
1,2021-08-05 16:34:17.726263-03:00,R$ 430.000,"Rua R 7, Setor Oeste",66 - 68 m²,2,2,2 - 3,NaN,NaN,apartamentos
2,2021-08-05 16:34:17.726628-03:00,R$ 345.000,"Rua 7, Setor Oeste",34 - 76 m²,1 - 2,1,1,NaN,NaN,apartamentos
3,2021-08-05 16:34:17.726966-03:00,R$ 420.000,"Setor Bueno, Goiânia",89 m²,3,1,2,NaN,NaN,apartamentos
4,2021-08-05 16:34:17.727328-03:00,R$ 1.018.000,"Setor Bueno, Goiânia",175 m²,4,5,4,R$ 650,R$ 2.000,apartamentos


5 últimas linhas:


,DATE,PRICE,ADDRESS,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU,TIPO
22105,2021-08-05 16:45:57.314919-03:00,R$ 2.500.000,"Rua T 27, Setor Bueno",750 m²,NaN,NaN,NaN,NaN,NaN,terrenos-lotes-condominios
22106,2021-08-05 16:45:57.315271-03:00,R$ 200.000,"Rua do Buriti, Bairro Santa Rita",420 m²,NaN,NaN,NaN,NaN,NaN,terrenos-lotes-condominios
22107,2021-08-05 16:45:57.315586-03:00,R$ 1.700.000,"Rodovia GO-060, Chácaras Maringá",10000 m²,NaN,NaN,NaN,NaN,R$ 1.400,terrenos-lotes-condominios
22108,2021-08-05 16:45:57.315956-03:00,R$ 5.500.000,"Avenida Americano do Brasil, Parque Santa Rita",10314 m²,NaN,NaN,NaN,NaN,NaN,terrenos-lotes-condominios
22109,2021-08-05 16:45:57.316281-03:00,R$ 2.000.000,"Rua do Salmão, Jardim Atlântico",2495 m²,NaN,NaN,NaN,NaN,NaN,terrenos-lotes-condominios


5 linhas aleatórias:


,DATE,PRICE,ADDRESS,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU,TIPO
12599,2021-08-05 19:32:44.069968-03:00,R$ 165.000,"Residencial Campos Dourados, Goiânia",150 m²,2,2,1,NaN,NaN,casas
9391,2021-08-05 16:34:21.936913-03:00,R$ 600.000,"Rua T 52, Setor Bueno",110 m²,3,2,4,NaN,NaN,apartamentos
5714,2021-08-05 16:34:20.290574-03:00,R$ 381.494,"Avenida Afonso Pena, Vila dos Alpes",65 m²,2,2,2,NaN,NaN,apartamentos
13463,2021-08-05 19:32:44.431844-03:00,R$ 550.000,"Jardim Atlântico, Goiânia",126 m²,3,4,2,NaN,NaN,casas
16762,2021-08-05 18:08:25.307440-03:00,R$ 1.200.000,"Rua T 36, Setor Bueno",303 m²,4,3,6,R$ 1.100,NaN,cobertura


**Primeiras Observacoes**
- O dataset contém inúmeros dados nulos/faltantes. Requer atencao para a tomada de decisao sobre como eles serao tratados.
- O dataset contem 22100 linhas e 10 colunas.
- Todas as colunas do dataset estao sendo consideradas como 'object'; isso impede o cálculo correto acerca das medidas dos dados.

**Diagnostico de Qualidade**
- Nessa etapa, serao identificados valores ausentes, duplicados, únicos, etc.

In [15]:
# Dados Nulos onde true = nulo; false = nao-nulo
gyn.isna()

,DATE,PRICE,ADDRESS,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU,TIPO
0,False,False,False,False,False,False,False,True,True,False
1,False,False,False,False,False,False,False,True,True,False
2,False,False,False,False,False,False,False,True,True,False
3,False,False,False,False,False,False,False,True,True,False
4,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...
22105,False,False,False,False,True,True,True,True,True,False
22106,False,False,False,False,True,True,True,True,True,False
22107,False,False,False,False,True,True,True,True,False,False
22108,False,False,False,False,True,True,True,True,True,False


In [16]:
# quantidades
quantidade_ausentes = gyn.isna().sum()
display(quantidade_ausentes)

,0
DATE,0
PRICE,0
ADDRESS,3
AREAS,38
BEDROOMS,1445
PARKING-SPACES,2640
BATHROOMS,1429
CONDOMÍNIO,10752
IPTU,15441
TIPO,0


In [17]:
# percentual de ausentes
percentual_ausentes = (gyn.isna().mean() * 100).round(2)
percentual_ausentes

,0
DATE,0.00
PRICE,0.00
ADDRESS,0.01
AREAS,0.17
BEDROOMS,6.54
PARKING-SPACES,11.94
BATHROOMS,6.46
CONDOMÍNIO,48.63
IPTU,69.84
TIPO,0.00


In [18]:
diagnostico = pd.DataFrame({
    'Tipo_Dado': gyn.dtypes,
    'Quantidade_Ausentes': quantidade_ausentes,
    'Percentual_Ausentes': percentual_ausentes,
    'Quantidade_Unicos': gyn.nunique()
})
diagnostico.index.name = 'Colunas'
diagnostico


,Tipo_Dado,Quantidade_Ausentes,Percentual_Ausentes,Quantidade_Unicos
Colunas,,,,
DATE,object,0,0.00,22110
PRICE,object,0,0.00,2135
ADDRESS,object,3,0.01,3931
AREAS,object,38,0.17,1087
BEDROOMS,object,1445,6.54,21
PARKING-SPACES,object,2640,11.94,35
BATHROOMS,object,1429,6.46,28
CONDOMÍNIO,object,10752,48.63,444
IPTU,object,15441,69.84,629


In [19]:
diagnostico_com_ausencias = (
    diagnostico.loc[
        diagnostico['Quantidade_Ausentes'] > 0,
    ].sort_values(
        by='Percentual_Ausentes',
        ascending=False
    )
)
diagnostico_com_ausencias


,Tipo_Dado,Quantidade_Ausentes,Percentual_Ausentes,Quantidade_Unicos
Colunas,,,,
IPTU,object,15441,69.84,629
CONDOMÍNIO,object,10752,48.63,444
PARKING-SPACES,object,2640,11.94,35
BEDROOMS,object,1445,6.54,21
BATHROOMS,object,1429,6.46,28
AREAS,object,38,0.17,1087
ADDRESS,object,3,0.01,3931


**Resumo dos Valores Encontrados**
- A coluna do IPTU tem a maior quantidade de dados ausentes (69.84%), seguida de condomínio (48.63%). Ou seja, ambas tem as maiores quantidades e percentuais de ausencia.


In [20]:
# Duplicatas
quant_duplicadas_completas = gyn.duplicated().sum()
mascara_duplicadas_completas = gyn.duplicated(keep=False)
registros_duplicados_completos = gyn.loc[
    mascara_duplicadas_completas
].sort_values(
    by='AREAS'
)

In [21]:
diagnostico.loc["IPTU"]

,IPTU
Tipo_Dado,object
Quantidade_Ausentes,15441
Percentual_Ausentes,69.84
Quantidade_Unicos,629


In [22]:
diagnostico.loc["PRICE"]


,PRICE
Tipo_Dado,object
Quantidade_Ausentes,0
Percentual_Ausentes,0.0
Quantidade_Unicos,2135


In [23]:
print("Duplicatas excedentes:",
      quant_duplicadas_completas)
print("Quantidade de linhas envoldidas:",
      mascara_duplicadas_completas.sum())
registros_duplicados_completos

# Nao possui linhas duplicatadas

Duplicatas excedentes: 0
Quantidade de linhas envoldidas: 0


Colunas,DATE,PRICE,ADDRESS,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU,TIPO


## **Conclusao do Diagnostico de Qualidade**
- A tabela do diagnostico de qualidade contem o tipo do dado, a quantidade de dados unicos, ausentes e o percentual de dados ausentes.
- Das 10 counas, `7` possuem valores ausentes.
- A coluna com o maior percentual de ausencia é `IPTU (69,84%)`, com `15.441` valores ausentes.
- O dataset nao apresenta dados repetidos/duplicados.
- Das 10 colunas, `8` possuem tipos inadequados. Apenas `ADDRESS` e `TIPO` possuem tipos adequados.
- Nenhum registro foi corrigido ou removido durante essa etapa.

**Limpeza de Coluna**
- Após uma análise acerca da coluna `DATE`, concluimos que ela nao possui utiliade no trabalho.
- Em razao, disso, ela sera excluida.

In [25]:
gyn.drop(columns=['DATE'], inplace=True)

KeyError: "['DATE'] not found in axis"

In [26]:
gyn.head()

Colunas,PRICE,ADDRESS,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU,TIPO
0,Sob consulta,"Rua João de Abreu, Setor Oeste",222 - 485 m²,3 - 4,2 - 4,3 - 5,NaN,NaN,apartamentos
1,R$ 430.000,"Rua R 7, Setor Oeste",66 - 68 m²,2,2,2 - 3,NaN,NaN,apartamentos
2,R$ 345.000,"Rua 7, Setor Oeste",34 - 76 m²,1 - 2,1,1,NaN,NaN,apartamentos
3,R$ 420.000,"Setor Bueno, Goiânia",89 m²,3,1,2,NaN,NaN,apartamentos
4,R$ 1.018.000,"Setor Bueno, Goiânia",175 m²,4,5,4,R$ 650,R$ 2.000,apartamentos


**Conversao de Tratamento de Dados**
- Quase todas as colunas possuem formatos errados e, por conta disso, serao tratados.

### Limpeza e Conversão da Coluna 'PRICE'

A coluna 'PRICE' contém valores monetários em formato de texto (`R$ 430.000`, `Sob consulta`).

**Passos:**
1.  Substituir 'Sob consulta' por `NaN`.
2.  Remover 'R$', '.' e substituir ',' por '.' para formatar como número.
3.  Converter a coluna para o tipo `float`.

In [28]:
# Limpeza e conversão da coluna 'PRICE'
gyn['PRICE'] = gyn['PRICE'].replace('Sob consulta', np.nan)  # Substituir 'Sob consulta' por NaN
gyn['PRICE'] = gyn['PRICE'].str.replace('R$', '', regex=False) # Remover 'R$'
gyn['PRICE'] = gyn['PRICE'].str.replace('.', '', regex=False)  # Remover separador de milhares '.'
gyn['PRICE'] = gyn['PRICE'].str.replace(',', '.', regex=False)  # Substituir ',' por '.' para decimais
gyn['PRICE'] = pd.to_numeric(gyn['PRICE'], errors='coerce') # Converter para numérico, forçando erros para NaN

print("Coluna 'PRICE' convertida:")
display(gyn[['PRICE']].head())

Coluna 'PRICE' convertida:


Colunas,PRICE
0,NaN
1,430000.0
2,345000.0
3,420000.0
4,1018000.0


### Limpeza e Conversão das Colunas 'CONDOMÍNIO' e 'IPTU'

As colunas 'CONDOMÍNIO' e 'IPTU' também contêm valores monetários em formato de texto.

**Passos:**
1.  Remover 'R$', '.' e substituir ',' por '.' para formatar como número.
2.  Converter as colunas para o tipo `float`.

In [29]:
# Limpeza e conversão da coluna 'CONDOMÍNIO'
gyn['CONDOMÍNIO'] = gyn['CONDOMÍNIO'].str.replace('R$', '', regex=False)
gyn['CONDOMÍNIO'] = gyn['CONDOMÍNIO'].str.replace('.', '', regex=False)
gyn['CONDOMÍNIO'] = gyn['CONDOMÍNIO'].str.replace(',', '.', regex=False)
gyn['CONDOMÍNIO'] = pd.to_numeric(gyn['CONDOMÍNIO'], errors='coerce')

# Limpeza e conversão da coluna 'IPTU'
gyn['IPTU'] = gyn['IPTU'].str.replace('R$', '', regex=False)
gyn['IPTU'] = gyn['IPTU'].str.replace('.', '', regex=False)
gyn['IPTU'] = gyn['IPTU'].str.replace(',', '.', regex=False)
gyn['IPTU'] = pd.to_numeric(gyn['IPTU'], errors='coerce')

print("Colunas 'CONDOMÍNIO' e 'IPTU' convertidas:")
display(gyn[['CONDOMÍNIO', 'IPTU']].head())

Colunas 'CONDOMÍNIO' e 'IPTU' convertidas:


Colunas,CONDOMÍNIO,IPTU
0,NaN,NaN
1,NaN,NaN
2,NaN,NaN
3,NaN,NaN
4,650.0,2000.0


### Limpeza e Conversão da Coluna 'AREAS'

A coluna 'AREAS' contém strings como `222 - 485 m²` ou `66 m²`.

**Passos:**
1.  Remover ' m²' e espaços em branco.
2.  Para ranges (ex: '222 - 485'), extrair o primeiro número (inferior).
3.  Converter a coluna para o tipo `float`.

In [30]:
# Limpeza e conversão da coluna 'AREAS'
# Remover ' m²' e extrair o primeiro número de um possível range
gyn['AREAS'] = gyn['AREAS'].str.replace(' m²', '', regex=False).str.strip()
gyn['AREAS'] = gyn['AREAS'].apply(lambda x: str(x).split(' ')[0] if pd.notna(x) else np.nan)
gyn['AREAS'] = pd.to_numeric(gyn['AREAS'], errors='coerce')

print("Coluna 'AREAS' convertida:")
display(gyn[['AREAS']].head())

Coluna 'AREAS' convertida:


Colunas,AREAS
0,222.0
1,66.0
2,34.0
3,89.0
4,175.0


### Limpeza e Conversão das Colunas 'BEDROOMS', 'PARKING-SPACES', 'BATHROOMS'

Essas colunas contêm strings que podem ser números únicos ou ranges (ex: `3 - 4`).

**Passos:**
1.  Para ranges, extrair o primeiro número (inferior).
2.  Converter as colunas para o tipo `float` (para lidar com `NaN`s) e depois para `Int64` (tipo inteiro que aceita `NaN`).

In [31]:
# Funções auxiliares para extrair o primeiro número de um range e converter para numérico
def extract_first_number(value):
    if pd.isna(value):
        return np.nan
    s_value = str(value).strip()
    if '-' in s_value:
        return pd.to_numeric(s_value.split('-')[0].strip(), errors='coerce')
    return pd.to_numeric(s_value, errors='coerce')

columns_to_convert_int = ['BEDROOMS', 'PARKING-SPACES', 'BATHROOMS']

for col in columns_to_convert_int:
    gyn[col] = gyn[col].apply(extract_first_number)
    # Converter para float primeiro para manter NaN, depois para Int64
    gyn[col] = gyn[col].astype('float').astype(pd.Int64Dtype())

print("Colunas de quantidade convertidas:")
display(gyn[columns_to_convert_int].head())

Colunas de quantidade convertidas:


Colunas,BEDROOMS,PARKING-SPACES,BATHROOMS
0,3,2,3
1,2,2,2
2,1,1,1
3,3,1,2
4,4,5,4


### Verificação Final dos Tipos de Dados

Agora vamos verificar os tipos de dados de todas as colunas para garantir que as conversões foram aplicadas corretamente.

In [32]:
# Verificar os tipos de dados atualizados
print("Informações atualizadas do DataFrame:")
gyn.info()

print("Primeiras 5 linhas do DataFrame após a limpeza:")
display(gyn.head())

Informações atualizadas do DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22110 entries, 0 to 22109
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   PRICE           22086 non-null  float64
 1   ADDRESS         22107 non-null  object 
 2   AREAS           22072 non-null  float64
 3   BEDROOMS        20665 non-null  Int64  
 4   PARKING-SPACES  19470 non-null  Int64  
 5   BATHROOMS       20681 non-null  Int64  
 6   CONDOMÍNIO      11358 non-null  float64
 7   IPTU            6669 non-null   float64
 8   TIPO            22110 non-null  object 
dtypes: Int64(3), float64(4), object(2)
memory usage: 1.6+ MB
Primeiras 5 linhas do DataFrame após a limpeza:


Colunas,PRICE,ADDRESS,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU,TIPO
0,NaN,"Rua João de Abreu, Setor Oeste",222.0,3,2,3,NaN,NaN,apartamentos
1,430000.0,"Rua R 7, Setor Oeste",66.0,2,2,2,NaN,NaN,apartamentos
2,345000.0,"Rua 7, Setor Oeste",34.0,1,1,1,NaN,NaN,apartamentos
3,420000.0,"Setor Bueno, Goiânia",89.0,3,1,2,NaN,NaN,apartamentos
4,1018000.0,"Setor Bueno, Goiânia",175.0,4,5,4,650.0,2000.0,apartamentos


In [33]:
gyn.describe()

Colunas,PRICE,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU
count,2.208600e+04,2.207200e+04,20665.0,19470.0,20681.0,1.135800e+04,6669.000000
mean,1.226318e+06,3.603652e+04,3.085604,2.583102,3.355399,2.332898e+03,1121.684660
std,1.412380e+07,4.066242e+06,0.919462,3.441214,1.768763,5.291340e+04,9906.436859
min,1.000000e+03,1.000000e+00,1.0,0.0,1.0,1.000000e+00,1.000000
25%,3.200000e+05,8.500000e+01,3.0,2.0,2.0,3.500000e+02,331.000000
50%,5.500000e+05,1.310000e+02,3.0,2.0,3.0,5.000000e+02,700.000000
75%,1.166000e+06,2.850000e+02,4.0,4.0,5.0,6.700000e+02,1299.000000
max,1.300000e+09,5.700000e+08,30.0,421.0,31.0,3.500000e+06,790000.000000


### Análise de Média e Mediana para Imputação de Valores Ausentes

Antes de preencher os valores ausentes, vamos calcular a média e a mediana para as colunas numéricas que ainda contêm `NaN`s. A comparação entre esses dois valores nos ajuda a entender a simetria da distribuição dos dados:
- Se a média e a mediana são próximas, a distribuição é mais simétrica e a média pode ser uma boa opção.
- Se há uma diferença significativa, indicando uma distribuição assimétrica (enviesada), a mediana é geralmente uma escolha mais robusta para evitar distorções causadas por *outliers*.

In [34]:
cols_to_impute_numerical = [
    'PRICE', 'AREAS', 'BEDROOMS', 'PARKING-SPACES', 'BATHROOMS', 'CONDOMÍNIO', 'IPTU'
]

# Criar um DataFrame para armazenar as estatísticas
stats_df = pd.DataFrame(index=cols_to_impute_numerical, columns=['Média', 'Mediana'])

print("Média e Mediana das colunas numéricas com valores ausentes:")
for col in cols_to_impute_numerical:
    if gyn[col].isnull().any(): # Apenas se houver valores ausentes
        mean_val = gyn[col].mean()
        median_val = gyn[col].median()
        stats_df.loc[col, 'Média'] = mean_val
        stats_df.loc[col, 'Mediana'] = median_val
    else:
        stats_df.loc[col, 'Média'] = 'N/A (sem NaN)'
        stats_df.loc[col, 'Mediana'] = 'N/A (sem NaN)'

display(stats_df.dropna(how='all')) # Exibe apenas as colunas que realmente precisavam de cálculo


Média e Mediana das colunas numéricas com valores ausentes:


,Média,Mediana
PRICE,1226317.667527,550000.0
AREAS,36036.515857,131.0
BEDROOMS,3.085604,3.0
PARKING-SPACES,2.583102,2.0
BATHROOMS,3.355399,3.0
CONDOMÍNIO,2332.897781,500.0
IPTU,1121.68466,700.0


### Estratégia de Imputação de Valores Ausentes

Com base na análise de média e mediana e na sua preferência pela mediana em casos de distribuições assimétricas, usaremos a seguinte estratégia:

*   Para as colunas **`PRICE`, `AREAS`, `CONDOMÍNIO`, `IPTU`, `BEDROOMS`, `PARKING-SPACES`, `BATHROOMS`**, onde a mediana é mais robusta contra *outliers* do que a média, preencheremos os valores ausentes com a **mediana**.
*   Para a coluna **`ADDRESS`**, que é um dado textual e a média/mediana não se aplicam, ela será mantida com os valores ausentes por enquanto, pois a imputação simples com a moda pode não ser adequada para seu contexto (endereços).


In [35]:
# Imputando valores ausentes com a mediana para as colunas numéricas
for col in cols_to_impute_numerical:
    if gyn[col].isnull().any():
        median_val = gyn[col].median()
        gyn[col].fillna(median_val, inplace=True)
        print(f"Valores ausentes da coluna '{col}' preenchidos com a mediana: {median_val}")

print("\nVerificando valores ausentes após a imputação:")
display(gyn.isnull().sum())

print("\nInformações atualizadas do DataFrame após a imputação:")
gyn.info()

print("\nPrimeiras 5 linhas do DataFrame após a imputação:")
display(gyn.head())


Valores ausentes da coluna 'PRICE' preenchidos com a mediana: 550000.0
Valores ausentes da coluna 'AREAS' preenchidos com a mediana: 131.0
Valores ausentes da coluna 'BEDROOMS' preenchidos com a mediana: 3.0
Valores ausentes da coluna 'PARKING-SPACES' preenchidos com a mediana: 2.0
Valores ausentes da coluna 'BATHROOMS' preenchidos com a mediana: 3.0
Valores ausentes da coluna 'CONDOMÍNIO' preenchidos com a mediana: 500.0
Valores ausentes da coluna 'IPTU' preenchidos com a mediana: 700.0

Verificando valores ausentes após a imputação:


/tmp/ipykernel_1399/1285361006.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  gyn[col].fillna(median_val, inplace=True)
/tmp/ipykernel_1399/1285361006.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using '

,0
Colunas,
PRICE,0
ADDRESS,3
AREAS,0
BEDROOMS,0
PARKING-SPACES,0
BATHROOMS,0
CONDOMÍNIO,0
IPTU,0
TIPO,0



Informações atualizadas do DataFrame após a imputação:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22110 entries, 0 to 22109
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   PRICE           22110 non-null  float64
 1   ADDRESS         22107 non-null  object 
 2   AREAS           22110 non-null  float64
 3   BEDROOMS        22110 non-null  Int64  
 4   PARKING-SPACES  22110 non-null  Int64  
 5   BATHROOMS       22110 non-null  Int64  
 6   CONDOMÍNIO      22110 non-null  float64
 7   IPTU            22110 non-null  float64
 8   TIPO            22110 non-null  object 
dtypes: Int64(3), float64(4), object(2)
memory usage: 1.6+ MB

Primeiras 5 linhas do DataFrame após a imputação:


Colunas,PRICE,ADDRESS,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU,TIPO
0,550000.0,"Rua João de Abreu, Setor Oeste",222.0,3,2,3,500.0,700.0,apartamentos
1,430000.0,"Rua R 7, Setor Oeste",66.0,2,2,2,500.0,700.0,apartamentos
2,345000.0,"Rua 7, Setor Oeste",34.0,1,1,1,500.0,700.0,apartamentos
3,420000.0,"Setor Bueno, Goiânia",89.0,3,1,2,500.0,700.0,apartamentos
4,1018000.0,"Setor Bueno, Goiânia",175.0,4,5,4,650.0,2000.0,apartamentos


### Localizando Linhas com Endereços Ausentes

Agora vamos identificar e exibir as 3 linhas onde a coluna `ADDRESS` ainda possui valores ausentes.

In [36]:
linhas_com_endereco_ausente = gyn[gyn['ADDRESS'].isna()]
print("Linhas do DataFrame com endereços ausentes:")
display(linhas_com_endereco_ausente)

Linhas do DataFrame com endereços ausentes:


Colunas,PRICE,ADDRESS,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU,TIPO
13112,280000.0,NaN,131.0,3,2,3,500.0,700.0,casas
14750,200000.0,NaN,131.0,3,2,3,500.0,700.0,casas
21724,500000.0,NaN,131.0,3,2,3,500.0,700.0,terrenos-lotes-condominios


### Excluindo Linhas com Endereços Ausentes

Conforme solicitado, removeremos as 3 linhas identificadas com valores `NaN` na coluna `ADDRESS`.

In [37]:
# Remover as linhas onde 'ADDRESS' é NaN
gyn.dropna(subset=['ADDRESS'], inplace=True)

print("Valores ausentes após a exclusão das linhas com ADDRESS nulo:")
display(gyn.isnull().sum())

print("Primeiras 5 linhas do DataFrame após a exclusão:")
display(gyn.head())

Valores ausentes após a exclusão das linhas com ADDRESS nulo:


,0
Colunas,
PRICE,0
ADDRESS,0
AREAS,0
BEDROOMS,0
PARKING-SPACES,0
BATHROOMS,0
CONDOMÍNIO,0
IPTU,0
TIPO,0


Primeiras 5 linhas do DataFrame após a exclusão:


Colunas,PRICE,ADDRESS,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU,TIPO
0,550000.0,"Rua João de Abreu, Setor Oeste",222.0,3,2,3,500.0,700.0,apartamentos
1,430000.0,"Rua R 7, Setor Oeste",66.0,2,2,2,500.0,700.0,apartamentos
2,345000.0,"Rua 7, Setor Oeste",34.0,1,1,1,500.0,700.0,apartamentos
3,420000.0,"Setor Bueno, Goiânia",89.0,3,1,2,500.0,700.0,apartamentos
4,1018000.0,"Setor Bueno, Goiânia",175.0,4,5,4,650.0,2000.0,apartamentos


In [38]:
# Chegagem de informacoes
gyn.info()

<class 'pandas.core.frame.DataFrame'>
Index: 22107 entries, 0 to 22109
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   PRICE           22107 non-null  float64
 1   ADDRESS         22107 non-null  object 
 2   AREAS           22107 non-null  float64
 3   BEDROOMS        22107 non-null  Int64  
 4   PARKING-SPACES  22107 non-null  Int64  
 5   BATHROOMS       22107 non-null  Int64  
 6   CONDOMÍNIO      22107 non-null  float64
 7   IPTU            22107 non-null  float64
 8   TIPO            22107 non-null  object 
dtypes: Int64(3), float64(4), object(2)
memory usage: 1.7+ MB


In [39]:
gyn.describe()

Colunas,PRICE,AREAS,BEDROOMS,PARKING-SPACES,BATHROOMS,CONDOMÍNIO,IPTU
count,2.210700e+04,2.210700e+04,22107.0,22107.0,22107.0,2.210700e+04,22107.000000
mean,1.225706e+06,3.597967e+04,3.08002,2.513548,3.332474,1.441695e+03,827.209255
std,1.411711e+07,4.063022e+06,0.889219,3.234975,1.712989,3.793753e+04,5444.209358
min,1.000000e+03,1.000000e+00,1.0,0.0,1.0,1.000000e+00,1.000000
25%,3.200000e+05,8.500000e+01,3.0,2.0,2.0,4.900000e+02,700.000000
50%,5.500000e+05,1.310000e+02,3.0,2.0,3.0,5.000000e+02,700.000000
75%,1.163465e+06,2.840000e+02,3.0,3.0,4.0,5.000000e+02,700.000000
max,1.300000e+09,5.700000e+08,30.0,421.0,31.0,3.500000e+06,790000.000000


In [41]:
gyn.to_excel('EDA - Cabral.xlsx', index=False)